crates/ripemd160/src/ripemd160.rs

```diff
@@ -0,0 +1,473 @@
+//! Circuits for the [RIPEMD-160] hash function and its internal compression
+//! function.
+//!
+//! [RIPEMD-160]: https://www.esat.kuleuven.be/cosic/publications/article-317.pdf
+
+#![allow(clippy::many_single_char_names)]
+
+use ff::PrimeField;
+
+use super::uint32::UInt32;
+use bellpepper::gadgets::boolean::Boolean;
+use bellpepper::gadgets::multieq::MultiEq;
+use bellpepper_core::{ConstraintSystem, SynthesisError};
+
+#[allow(clippy::unreadable_literal)]
+const MD_BUFFERS: [u32; 5] = [0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0];
+
+#[allow(clippy::unreadable_literal)]
+const MD_BUFFERS_PRIME: [u32; 5] = [0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0];
+
+#[allow(clippy::unreadable_literal)]
+const K_BUFFER: [u32; 5] = [0x00000000, 0x5a827999, 0x6ed9eba1, 0x8f1bbcdc, 0xa953fd4e];
+
+#[allow(clippy::unreadable_literal)]
+const K_BUFFER_PRIME: [u32; 5] = [0x50a28be6, 0x5c4dd124, 0x6d703ef3, 0x7a6d76e9, 0x00000000];
+
+pub fn ripemd160<Scalar, CS>(
+    mut cs: CS,
+    input: &[Boolean],
+) -> Result<Vec<Boolean>, bellpepper_core::SynthesisError>
+where
+    Scalar: PrimeField,
+    CS: ConstraintSystem<Scalar>,
+{
+    assert!(input.len() % 8 == 0);
+
+    let mut padded = input.to_vec();
+    let plen = padded.len() as u64;
+    padded.push(Boolean::Constant(true));
+    while (padded.len() + 64) % 512 != 0 {
+        padded.push(Boolean::constant(false));
+    }
+    for b in (0..64).rev().map(|i| (plen >> i) & 1 == 1) {
+        padded.push(Boolean::constant(b));
+    }
+    assert!(padded.len() % 512 == 0);   
+    let mut cur_md = get_ripemd160_md("md");
+    let mut cur_md_prime = get_ripemd160_md("md_prime");
+    for (i, block) in padded.chunks(512).enumerate() {
+        let prev_md = cur_md.clone();
+        cur_md = ripemd160_func_block(cs.namespace(|| format!("block {}", i)), block, &mut cur_md)?;
+        cur_md_prime = ripemd160_func_block_prime(
+            cs.namespace(|| format!("block_prime {}", i)),
+            block,
+            &mut cur_md_prime,
+        )?;
+        let mut update_md = cur_md.clone();
+        for i in 0..5 {
+            update_md[(i+4)%5] = prev_md[i].xor(
+                cs.namespace(|| format!("first xor {}", i)),
+                &cur_md[(i+1)%5],
+            )?;
+            update_md[(i+4)%5] = update_md[(i+4)%5].xor(
+                cs.namespace(|| format!("second xor {}", i)),
+                &cur_md_prime[(i+2)%5],
+            )?;
+        }
+        cur_md = update_md;
+        cur_md_prime = cur_md.clone();
+    }
+    Ok(cur_md.into_iter().flat_map(|e| e.into_bits_be()).collect())
+}
+
+fn get_ripemd160_md(input: &str) -> Vec<UInt32> {
+    match input {
+        "md" => MD_BUFFERS.iter().map(|&v| UInt32::constant(v)).collect(),
+        "md_prime" => MD_BUFFERS_PRIME
+            .iter()
+            .map(|&v| UInt32::constant(v))
+            .collect(),
+        _ => panic!("Invalid input"),
+    }
+}
+
+pub fn ripemd160_func_block<Scalar, CS>(
```

Comment:
The implementation of this function seems to unroll the RIPEMD160 loop, leading to verbose code. This can be shortened. See the non-circuit implementation [here](https://github.com/avras/bitcoinvm/blob/2d6d6bd7f29a4438eeec9a483c56043ddcb81c6a/gadgets/src/ripemd160/ref_impl/ripemd160.rs#L90).

Also, including `left` and `right` in the names of the functions would make it easier to read.

Reply:
Thanks for the ref link. It is really helpful. Will modify the code accordingly.

Reply:
Modified the code and made it less verbose.